Support Test Impact Analysis

mpdeimos opened the pull request:
Addresses issue [TS-38029](https://jira.cqse.eu/browse/TS-38029)

- [x] Tests written
- [x] Documentation updated
- [x] `CHANGELOG.md` updated
- [ ] Present new features in [N&N](https://wiki.cqse.eu/pages/viewpage.action?pageId=689566)

Please respect the [Teamscale](https://demo.teamscale.com/activity.html#/teamscale-net-profiler) vote. Clean up findings before sending your change into review. Blacklist and report false-positives. Adjust the analysis profile if needed.

mpdeimos committed:
zmq powered gui

mpdeimos committed:
retarget vc projects

mpdeimos committed:
codemaid

mpdeimos committed:
add ZMQ, first concept of one-way profiler communication

mpdeimos committed:
add testcase for one-way profiler communication

mpdeimos committed:
rename gui

mpdeimos committed:
add profiler commons project and use in test

mpdeimos committed:
Test infra: use .net ZMQ and native ZMQ

salsolatragus committed:
Wrap coverage data in report class.

mpdeimos committed:
Extend Profiler Testee with an interactive mode

mpdeimos committed:
Add publisher socket to C# IPC implementation

mpdeimos committed:
Wire subscribe socket in C++ IPC

mpdeimos committed:
write function infos to log on test case switch

salsolatragus committed:
Write testwise coverage report.

mpdeimos committed:
Real TIA method call profiling

mpdeimos committed:
Refactor tests

mpdeimos committed:
No locking until IPC is started

mpdeimos committed:
Use NetMQ-native threading

mpdeimos committed:
compiler profiler testee with native bitness and as x86, x64

mpdeimos committed:
parameterize tia tests as x64 and x86

mpdeimos committed:
test start ipc after app (not working yet)

mpdeimos committed:
commander rework

mpdeimos committed:
simplified commander

mpdeimos committed:
multipart tia start/end

mpdeimos committed:
escaping of characters

mpdeimos committed:
store test execution result

mpdeimos committed:
disable mscorlib profiling for TIA

mpdeimos committed:
fix for default socket

salsolatragus committed:
Merge TraceFile and ParsedTraceFile classes.

salsolatragus committed:
Hide internal data of TraceFile.

salsolatragus committed:
Make names consistent.

salsolatragus committed:
Ignores method references from unkown assemblies.

salsolatragus committed:
Improve some names.

salsolatragus committed:
Split testwise report and separate conversion from files.

salsolatragus committed:
Generate aggregated report.

salsolatragus committed:
Introduce ICoverageReport abstraction.

salsolatragus committed:
Tests simple report output.

salsolatragus committed:
Tests aggregated trace.

salsolatragus committed:
Test testwise report contents.

salsolatragus committed:
Remove obsolete method.

salsolatragus committed:
Simplify tests.

salsolatragus committed:
Test generating testwise report.

salsolatragus committed:
Merge SimpleCoverageReport and LineCoverageReport.

salsolatragus committed:
Remove unused import.

salsolatragus committed:
Remove obsolete class.

salsolatragus committed:
Cleanup testwise report model.

salsolatragus committed:
Simplify creation of testwise report.

salsolatragus committed:
Merge testwise reports with different tests.

salsolatragus committed:
Merge testwise reports.

salsolatragus committed:
Send report format to Teamscale.

salsolatragus committed:
Move simple report to own namespace.

salsolatragus committed:
Serialize testwise report.

salsolatragus committed:
Test parsing jitted, inlined, and called events.

salsolatragus committed:
Assume tests always start.

salsolatragus committed:
Consider test without end.

salsolatragus committed:
Add no-test coverage only if any.

salsolatragus committed:
Fix test result labels.

salsolatragus committed:
Keep non-test coverage.

mpdeimos committed:
cache latest test runs

salsolatragus committed:
Merge test metadata.

salsolatragus committed:
Simplify archiving API.

salsolatragus committed:
Shorten an expression.

salsolatragus committed:
Fix stupid merge error.

mpdeimos committed:
test timing fix

mpdeimos committed:
support test path prefixes

mpdeimos committed:
correct emptiness handling for testwise traces

mpdeimos committed:
test name validation

mpdeimos committed:
error logging

salsolatragus committed:
Extract Tia test baseclass.

salsolatragus committed:
Abstract test setup.

salsolatragus committed:
Restructure test infrastructure.

salsolatragus committed:
Remove TiaTestee.

salsolatragus committed:
Simplify testee.

salsolatragus committed:
Reduce visibility.

salsolatragus committed:
Specify bitness on testee rather than profiler.

salsolatragus committed:
Add performance testee.

salsolatragus committed:
Configure Commons to build a nuget package.

salsolatragus committed:
Remove unused import.

salsolatragus committed:
Add logging to commander.

salsolatragus committed:
Merge remote-tracking branch 'origin/tia_report' into tia

mpdeimos committed:
Performance improvements:
* use icorprofiler v3 interface
* only recod method calls if testcase is active
* latecall evaluation
* improved synchronization

mpdeimos committed:
Merge branch 'tia' of https://github.com/cqse/teamscale-profiler-dotnet into tia

salsolatragus committed:
Add missing dependency to commander.

salsolatragus commented on Profiler/utils/Ipc.cpp:
```diff
@@ -0,0 +1,145 @@
+
+	this->handlerThread = new std::thread(&Ipc::handlerThreadLoop, this);
+
+	this->request("profiler_connected");
```
Extract string to constant?

Raphael-N replied:
Done

salsolatragus commented on Profiler/utils/Ipc.cpp:
```diff
@@ -0,0 +1,145 @@
+		this->handlerThread->join();
+	}
+
+	this->request("profiler_disconnected");
```
Extract string to constant?

Raphael-N replied:
Done

salsolatragus commented on Profiler/utils/Ipc.cpp:
```diff
@@ -0,0 +1,145 @@
+
+void Ipc::handlerThreadLoop() {
+	this->zmqSubscribeSocket = zmq_socket(this->zmqContext, ZMQ_SUB);
+	zmq_setsockopt(this->zmqSubscribeSocket, ZMQ_SUBSCRIBE, "test:", 0);
```
Extract string to constant?

Raphael-N replied:
Done

salsolatragus commented on Profiler/utils/Ipc.cpp:
```diff
@@ -0,0 +1,145 @@
+}
+
+void Ipc::handleMessage(std::vector<std::string> frames) {
+	if (frames[0] == "test:start" && frames.size() > 1) {
```
Extract string to constant?

salsolatragus commented on Profiler/utils/Ipc.cpp:
```diff
@@ -0,0 +1,145 @@
+	if (frames[0] == "test:start" && frames.size() > 1) {
+		this->testStartCallback(frames[1]);
+	}
+	else if (frames[0] == "test:end") {
```
Extract string to constant?

Raphael-N replied:
Done

salsolatragus commented on Profiler/utils/Ipc.cpp:
```diff
@@ -0,0 +1,145 @@
+
+std::string Ipc::getCurrentTestName()
+{
+	return this->request("get_testname");
```
Extract string to constant?

Raphael-N replied:
Done

salsolatragus commented on Profiler/utils/Ipc.cpp:
```diff
@@ -0,0 +1,145 @@
+// ZMQ strings are not zero-terminated, let's terminate these
+#define IPC_TERMINATE_STRING(buffer, len) buffer[len < IPC_BUFFER_SIZE ? len : IPC_BUFFER_SIZE - 1] = '\0'
+
+Ipc::Ipc(Config* config, std::function<void(std::string)> testStartCallback, std::function<void(std::string, std::string)> testEndCallback, std::function<void(std::string)> errorCallback)
```
There's a lot of magic happening in this class, w.r.t the number of `frames` and `len` of some return value. This is very hard to follow without a deeper knowledge of ZMQ. Maybe some explanatory comments would ease maintenance?

salsolatragus commented on Profiler/utils/Ipc.cpp:
```diff
@@ -0,0 +1,145 @@
+
+#include <zmq.h>
+
+#define IPC_TIMEOUT_MS 250
```
Where does this number come from? Maybe this is the culprit in the distributed setting, because the profiler just doesn't wait long enough for the request to return?

Raphael-N replied:
Made this option configurable.

salsolatragus commented on Profiler_Test/Proxies/TiaProfiler.cs:
```diff
@@ -0,0 +1,134 @@
+                get
+                {
+                    FileInfo actualTrace = GetSingleTraceFile();
+                    TiaTestResult testResult = new TiaTestResult(File.ReadAllLines(actualTrace.FullName));
```
Should we reuse the logic for parsing trace files from the upload daemon here? It seems a bad idea to duplicate that parsing logic, even if in the tests.

salsolatragus commented on UploadDaemon/Scanning/TraceFile.cs:
```diff
@@ -42,18 +54,124 @@ public TraceFile(string filePath, string[] lines)
             return matchingLine?.Groups[1]?.Value;
         }
 
+        public ICoverageReport ToReport(Func<Trace, SimpleCoverageReport> traceResolver)
```
This method is a beast... should we introduce a builder to split this up?

salsolatragus reviewed (changes requested):
I went through the changes that you did, good work! I certainly learned a lot about how this works, though the ZMQ details are still a mystery to me 😅 

Here's some thoughts on what we should do:

* Do we want to stay backwards compatibility with the old profiler? I think it will hurt us sooner rather than later that we're dragging around `CorProfileV2` support.
* Conduct performance tests to get rid of some alternative implementations (and the respective ifdefs...)
* You review my work on the upload daemon and profiler tests.
* Cleanup trace file parsing in the upload daemon.
* Analyze whether the ZMQ connection timeout in the profiler is responsible for the distributed scenario to fail.
* Log in the profiler if connection to a commander could not be established due to timeout.

salsolatragus committed:
Add some interface comments.

salsolatragus committed:
Reduce default log level to Info.

mpdeimos committed:
fix empty traces

mpdeimos committed:
* keep existing tests also for report merge
* keep data on teamscale for existing tests

Raphael-N committed:
Add improvements to branch

Raphael-N committed:
Add comments and fix Testee bitness

mpdeimos committed:
remove support for known test cases

mpdeimos committed:
add support for partial coverage reports

Raphael-N committed:
Potential Fix for crashes at Healthineers

mpdeimos committed:
Fix casing

mpdeimos committed:
Commander CLI

mpdeimos committed:
Commander Server

mpdeimos committed:
Review

mpdeimos committed:
review

mpdeimos committed:
add logging

mpdeimos committed:
add quiet mode

mpdeimos committed:
additional null checks

mpdeimos committed:
add nullable library

mpdeimos committed:
forbid setting test name to null

mpdeimos committed:
🤦

mpdeimos committed:
better error handling

Raphael-N committed:
Merge pull request #160 from cqse/tia_server_cli

Tia Server & CLI

Raphael-N committed:
Update Platform version and toolset

Raphael-N committed:
Remove ifdefs, rework, add tga toggle to options

Raphael-N committed:
Swap vectors instead of using a toggle boolean

Raphael-N committed:
Rename method id vectors to primary and backing vectors

Raphael-N committed:
Merge pull request #155 from cqse/tia-merge-reports

Partial Reports & Improved Merging

Raphael-N committed:
Add tests for tga and tia configurations

Raphael-N committed:
Implement int set to achieve faster checks

Raphael-N committed:
Add documentation to custom set implementation

Raphael-N committed:
Merge branch 'tia' into performance_improvements_tia

Raphael-N committed:
Fix minor issue

Raphael-N committed:
Rework

mpdeimos committed:
Merge pull request #158 from cqse/performance_improvements_tia

Improve testwise coverage profiling performance by custom set implementation

Raphael-N committed:
Fix issue in functionId_set

mpdeimos committed:
update bummer/update output path

mpdeimos committed:
bummer update

mpdeimos committed:
fix tests

Raphael-N committed:
Use newer libzmq commands for context shutdown and add timeout to subscribe socket

mpdeimos committed:
pre-initialize profiler ipc

Raphael-N committed:
Add documentation for TIA mode catch in shutdown

Raphael-N committed:
Merge branch 'master' into tia

Raphael-N committed:
Fix Test cases

Raphael-N commented on Profiler/CProfilerCallback.cpp:
```diff
@@ -1,481 +1,667 @@
+}
+
+#ifdef TIA
+// TODO move to instance vars.
```
all `#ifdef TIA` removed

Raphael-N commented on Profiler/CProfilerCallback.cpp:
```diff
@@ -1,481 +1,667 @@
+		dumpEnvironment();
+	}
+
+#ifdef CorProfileV3
```
Got rid of CorProfilerV2

Raphael-N commented on Profiler/utils/Ipc.cpp:
```diff
@@ -0,0 +1,145 @@
+		return;
+	}
+
+	/// Alternative way to read messages
```
Done

Raphael-N commented on Profiler/log/TraceLog.cpp:
```diff
@@ -77,11 +84,41 @@ void TraceLog::logAssembly(std::string assembly)
+}
+#endif
+
+inline std::string TraceLog::escape(std::string message) {
```
Done

Raphael-N commented on Profiler/log/TraceLog.h:
```diff
@@ -63,6 +73,14 @@ class TraceLog: public FileLogBase
+#endif
+
+	/** The key to log information about test cases. */
+	const char* LOG_KEY_TESTCASE = "Test";
```
Removed ifdef TIA statements in favor of configuration.

Raphael-N commented on Profiler/CProfilerCallback.cpp:
```diff
@@ -1,481 +1,667 @@
+
+#ifdef LateCallEval
+		if (calledMethodIds.find(functionId) == calledMethodIds.end()) {
+			// TODO (MP) compare performance of mutexes
```
Tried it and mutexes performed worse 😄

Raphael-N commented on Profiler/CProfilerCallback.cpp:
```diff
@@ -1,481 +1,667 @@
+	jittedMethods.clear();
+
+#if TIA
+#ifdef LateCallEval
```
Chose to use the LateCallEval only since performance at method enter event turned out to be most crucial.

Raphael-N committed:
Consider testwise coverage for artifactory uploads

Raphael-N committed:
Change name for artifactory report upload

Raphael-N committed:
Merge branch 'master' into tia

Raphael-N committed:
Fix Test after merge

Raphael-N committed:
Remove Worker thread due to issues with many processes

Raphael-N committed:
Add Zipping step to Artifactory Upload

Raphael-N committed:
Fix Artifactory Upload

Raphael-N committed:
Improve artifactory branches

invalid-email-address committed:
add dialog for recording test duration

invalid-email-address committed:
implement long <-> string conversion for duration

invalid-email-address committed:
send adjusted duration to profiler

invalid-email-address committed:
fix bugs

invalid-email-address committed:
fix teamscale findings

Raphael-N committed:
Start implementing synchronous communication

Raphael-N committed:
Add Socket handling

Raphael-N committed:
Support time configuration in profiler and daemon

Raphael-N committed:
Use milliseconds for automated tests

Raphael-N committed:
Fix some communication issues

invalid-email-address committed:
rework

invalid-email-address committed:
update to newer framework version

invalid-email-address committed:
malformed trace lines should not crash upload daemon

Commit:
Send PID to ZmqIpcServer as clientID

Commit:
Update TargetFrameworkVersion in UploadDaemon_Test

Raphael-N committed:
rework

Raphael-N committed:
Rework

Raphael-N committed:
Write as one string instead of per entry

Raphael-N committed:
Add synchronization to jit inlining events

Raphael-N committed:
Fix casing of results

Raphael-N committed:
Merge branch 'tia_synchronous_events' of https://github.com/cqse/teamscale-profiler-dotnet into tia_synchronous_events

Raphael-N committed:
Disable inlining for accurate results

Raphael-N committed:
Merge pull request #180 from cqse/test-commander-change-duration

Test commander change duration

Raphael-N committed:
Fix encoding of test names

Raphael-N committed:
Update Test Times and Add Checks for running Tests

Raphael-N committed:
Fix initial test not appearing

Raphael-N committed:
Fix Process not shutting down

Raphael-N committed:
Improve performance of Upload Daemon

Raphael-N committed:
TS#35530 Check if the trace directory is writable

Raphael-N committed:
Fix issue with time sending

Raphael-N committed:
Reduce Memory Usage

Raphael-N committed:
Replace newline with space in testname

Raphael-N committed:
Implement Splitting for too large reports

Raphael-N committed:
Merge branch 'tia' into tia_synchronous_events

Raphael-N committed:
Cleanup leftovers from merge

Raphael-N committed:
Minor fixes

Raphael-N committed:
Fix several Teamscale findings

Raphael-N committed:
Add String report splitting for other upload types

Raphael-N committed:
Small fixes

Raphael-N committed:
Fix tests and socket configurations

Raphael-N committed:
Fix Tests

Raphael-N committed:
Update tests

Raphael-N committed:
More Test fixes and add back pipeline

Raphael-N committed:
Merge pull request #183 from cqse/tia_synchronous_events

Tia synchronous events

Raphael-N committed:
Merge branch 'master' into tia

Raphael-N committed:
Merge branch 'tia' of https://github.com/cqse/teamscale-profiler-dotnet into tia

Raphael-N committed:
Fix merge leftovers

Raphael-N committed:
TS#38029 Fix remaining tests

Raphael-N committed:
Merge branch 'master' into tia

Raphael-N committed:
TS#38029 Update Profiler Testee

Raphael-N committed:
TS#38029 Introduce AssemblyExtractor to fix EmbeddudUploadTarget Handling

Raphael-N committed:
TS#38029 Fix TS findings

Raphael-N committed:
TS#38029 Rework functionId_set

Raphael-N committed:
TS#38029 Refactor trace file parsing

Raphael-N committed:
TS#38029 Fix leftover issues from TraceFileParser rework

snowbagoly committed:
TS#38029 Delete Commander.CLI and wwwroot from Commander.Server

snowbagoly committed:
TS#38029 Delete Jacoco-based API

snowbagoly committed:
TS#38029 Improve UX of TestDurationDialog

snowbagoly committed:
TS#38029 Rework

snowbagoly commented on Commons/Ipc/ZmqIpcServer.cs:
```diff
@@ -0,0 +1,53 @@
+            this.responseSocket.ReceiveReady += (s, e) =>
+            {
+                string message = responseSocket.ReceiveFrameString();
+                string response = this.requestHandler(message);
```
This subclassing stuff was deleted in the meantime, so resolved 😁

snowbagoly committed:
TS-38029 Add CommanderTests to pipeline

snowbagoly committed:
TS-38029 Rename MainWindowViewModel props

snowbagoly committed:
Merge pull request #195 from cqse/tia_commander

Rework Tia commander

snowbagoly committed:
TS-38029 Rework Commons

snowbagoly committed:
TS-38029 Fix build error

wilhelma commented on Profiler/CProfilerCallback.h:
```diff
@@ -1,157 +1,184 @@
+	int assemblyCounter = 1;
+
+	bool isTestCaseRecording = false;
+	CProfilerCallback* callbackInstance = NULL;
```
```suggestion
	CProfilerCallback* callbackInstance = nullptr;
```

wilhelma commented on Profiler/CProfilerCallback.h:
```diff
@@ -1,157 +1,184 @@
+	AttachLog attachLog;
+
+	/** Inter-process connection for TIA communication. null if not in TIA mode. */
+	Ipc* ipc = NULL;
```
```suggestion
	Ipc* ipc = nullptr;
```

wilhelma commented on Profiler/CProfilerCallback.h:
```diff
@@ -1,157 +1,184 @@
+	Ipc* ipc = NULL;
+
+	/** Callback that is being called when a testcase starts. */
+	void onTestStart(std::string testName);
```
Avoid copies
```suggestion
	void onTestStart(const std::string& testName);
```

wilhelma commented on Profiler/CProfilerCallback.h:
```diff
@@ -1,157 +1,184 @@
+	 * We use the vector to uniquely store the information about called methods.
+	 */
+	std::vector<FunctionInfo> calledMethods;
+private:
```
unnecessary
```suggestion
```

wilhelma commented on Profiler/CProfilerCallback.h:
```diff
@@ -1,157 +1,184 @@
+	void getAssemblyInfo(AssemblyID assemblyId, WCHAR* assemblyName, WCHAR* assemblyPath, ASSEMBLYMETADATA* moduleId);
+
+	/** Triggers eagerly writing of function infos to log. */
+	void recordFunctionInfo(std::vector<FunctionInfo>* list, FunctionID calleeId);
```
Prefer using references for passing vectors, to have more type safety.
Consider if parameters are in (only read), out (return value), or in/out (read/write/return), see [cpp core guidelines on parameter passing](https://isocpp.github.io/CppCoreGuidelines/CppCoreGuidelines#f15-prefer-simple-and-conventional-ways-of-passing-information)
```suggestion
	void recordFunctionInfo(std::vector<FunctionInfo>& list, FunctionID calleeId);
```

wilhelma commented on Profiler/CProfilerCallback.h:
```diff
@@ -1,157 +1,184 @@
+	void onTestStart(std::string testName);
+
+	/** Callback that is being called when a testcase ends. */
+	void onTestEnd(std::string result = "", std::string duration = "");
```
As far as I can see `result` and `duration` are read-only
```suggestion
	void onTestEnd(const std::string& result = "", const std::string& duration = "");
```

wilhelma commented on Profiler/CProfilerCallback.h:
```diff
@@ -1,157 +1,184 @@
+	HRESULT InitializeImplementation(IUnknown* pICorProfilerInfoUnk);
+
+	/** Logs a stack trace. May rethrow the caught exception. */
+	void handleException(std::string context);
```
```suggestion
	void handleException(const std::string& context);
```

wilhelma commented on Profiler/CProfilerCallback.cpp:
```diff
@@ -1,488 +1,555 @@
+ * Serializes access to the singleton profiler instance when trying to shut it down.
+ * This prevents race conditions that can result in deadlocks that freeze the profiled process.
+ */
+class ShutdownGuard {
```
Generally, why don't you use an appropriate namespace?
It is unusual to have classes on the global namespace level, see [Google C++ guidelines](https://google.github.io/styleguide/cppguide.html#Namespaces)
I would use a namespace like `profiler`

wilhelma replied:
Generic question... 
I don't see any usage of the `ShutdownGuard`, can we delete it?

wilhelma commented on Profiler/CProfilerCallback.cpp:
```diff
@@ -1,488 +1,555 @@
+		InitializeCriticalSection(&section);
+	}
+
+	virtual ~ShutdownGuard() {}
```
see [cpp core guidlines](https://isocpp.github.io/CppCoreGuidelines/CppCoreGuidelines#c80-use-default-if-you-have-to-be-explicit-about-using-the-default-semantics)
```suggestion
	~ShutdownGuard() = default;
```

wilhelma commented on Profiler/CProfilerCallback.cpp:
```diff
@@ -1,488 +1,555 @@
+	 */
+	void shutdownInstance(bool clrIsAvailable) {
+		EnterCriticalSection(&section);
+		if (instance != NULL) {
```
search for all usages of `NULL` and replace with `nullptr`

wilhelma commented on Profiler/CProfilerCallback.cpp:
```diff
@@ -1,488 +1,555 @@
+
+	if (config.isTiaEnabled()) {
+		traceLog.info("TIA enabled. REQ Socket: " + config.getTiaRequestSocket());
+		std::function<void(std::string)> testStartCallback = std::bind(&CProfilerCallback::onTestStart, this, std::placeholders::_1);
+		std::function<void(std::string, std::string)> testEndCallback = std::bind(&CProfilerCallback::onTestEnd, this, std::placeholders::_1, std::placeholders::_2);
+		std::function<void(std::string)> errorCallback = std::bind(&TraceLog::error, this->traceLog, std::placeholders::_1);
```
Don't use `std::bind` but lambdas for readability (and maybe performance):
```suggestion
std::function<void(std::string)> testStartCallback = [this](std::string testName) {
    this->onTestStart(testName);
};

std::function<void(std::string, std::string)> testEndCallback = [this](std::string result, std::string duration) {
    this->onTestEnd(result, duration);
};

std::function<void(std::string)> errorCallback = [this](std::string message) {
    this->traceLog.error(message);
};
```

wilhelma commented on Profiler/CProfilerCallback.cpp:
```diff
@@ -1,488 +1,555 @@
+		std::function<void(std::string)> testStartCallback = std::bind(&CProfilerCallback::onTestStart, this, std::placeholders::_1);
+		std::function<void(std::string, std::string)> testEndCallback = std::bind(&CProfilerCallback::onTestEnd, this, std::placeholders::_1, std::placeholders::_2);
+		std::function<void(std::string)> errorCallback = std::bind(&TraceLog::error, this->traceLog, std::placeholders::_1);
+		this->ipc = new Ipc(&this->config, testStartCallback, testEndCallback, errorCallback);
```
Use a unique pointer instead of a raw pointer:
```suggestion
		this->ipc = std::make_unique<Ipc>(&this->config, testStartCallback, testEndCallback, errorCallback);
```

wilhelma commented on Profiler/CProfilerCallback.cpp:
```diff
@@ -1,488 +1,555 @@
+		setCalledMethodsSet(&calledMethodIds);
+	}
+
+	char appPool[BUFFER_SIZE];
```
use std::array (general, for all occurrences of old c-style arrays):
```suggestion
	std::array<char, BUFFER_SIZE> appPool;
```

wilhelma commented on Profiler/CProfilerCallback.cpp:
```diff
@@ -1,488 +1,555 @@
+	}
+
+	char appPool[BUFFER_SIZE];
+	if (GetEnvironmentVariable("APP_POOL_ID", appPool, sizeof(appPool))) {
```
With `std::array` you don't need two parameters as you can use `appPool.size()`
```suggestion
	if (GetEnvironmentVariable("APP_POOL_ID", appPool)) {
```

wilhelma commented on Profiler/CProfilerCallback.cpp:
```diff
@@ -1,488 +1,555 @@
+
+	char appPool[BUFFER_SIZE];
+	if (GetEnvironmentVariable("APP_POOL_ID", appPool, sizeof(appPool))) {
+		std::string message = "IIS AppPool: ";
+		message += appPool;
```
```suggestion
		std::string message = "IIS AppPool: " + appPool;
```

wilhelma commented on Profiler/CProfilerCallback.cpp:
```diff
@@ -1,488 +1,555 @@
+		traceLog.info(message);
+	}
+
+	std::string message = "Command Line: ";
+	message += GetCommandLine();
```
```suggestion
	std::string message = "Command Line: " + GetCommandLine();
```

wilhelma commented on Profiler/CProfilerCallback.cpp:
```diff
@@ -1,488 +1,555 @@
+		dumpEnvironment();
+	}
+
+	HRESULT hr = pICorProfilerInfoUnkown->QueryInterface(IID_ICorProfilerInfo3, (LPVOID*)&profilerInfo);
```
Don't use C-style casting (for all occurrences):
```suggestion
	HRESULT hr = pICorProfilerInfoUnkown->QueryInterface(IID_ICorProfilerInfo3, static_cast<LPVOID*>&profilerInfo);
```

wilhelma commented on Profiler/CProfilerCallback.cpp:
```diff
@@ -1,488 +1,555 @@
+}
+
+void CProfilerCallback::dumpEnvironment() {
+	std::vector<std::string> environmentVariables = WindowsUtils::listEnvironmentVariables();
```
Unnecessary copy (I am not sure if this works)...
```suggestion
	const std::vector<std::string>& environmentVariables = WindowsUtils::listEnvironmentVariables();
```

wilhelma commented on Profiler/CProfilerCallback.cpp:
```diff
@@ -1,488 +1,555 @@
+		return;
+	}
+
+	for (size_t i = 0; i < environmentVariables.size(); i++)
+	{
+		traceLog.logEnvironmentVariable(environmentVariables.at(i));
+	}
```
```suggestion
	for (const std::string variable : environmentVariables)
	{
		traceLog.logEnvironmentVariable(variable);
	}
```

wilhelma commented on Profiler/CProfilerCallback.cpp:
```diff
@@ -1,488 +1,555 @@
+	config.load(configFile, WindowsUtils::getPathOfThisProcess(), configFileWasManuallySpecified);
+}
+
+UploadDaemon CProfilerCallback::createDaemon() {
```
can be static

wilhelma commented on Profiler/CProfilerCallback.cpp:
```diff
@@ -1,488 +1,555 @@
+	traceLog.shutdown();
+	attachLog.shutdown();
+
+	if (this->ipc != NULL) {
+		delete ipc;
+		ipc = NULL;
+	}
```
With `unique_ptr`
```suggestion
	if (ipc != nullptr) {
		ipc.release();
	}
```

wilhelma commented on Profiler/CProfilerCallback.cpp:
```diff
@@ -1,488 +1,555 @@
+	}
+}
+
+HRESULT CProfilerCallback::AssemblyLoadFinishedImplementation(AssemblyID assemblyId, HRESULT hrStatus) {
```
`hrStatus` is not used.. Is this intended?

Raphael-N replied:
I am not sure and I did not really consider changing this.
It could make sense to check if the assembly load was successful before we handle it but I don't know if that is relevant or what can happen for an assembly load to fail.

Raphael-N replied:
Remove Name

wilhelma commented on Profiler/CProfilerCallback.cpp:
```diff
@@ -1,488 +1,555 @@
+
+	LeaveCriticalSection(&callbackSynchronization);
+
+	// Log assembly load.
+	writtenChars += sprintf_s(assemblyInfo + writtenChars, BUFFER_SIZE - writtenChars, "%S:%i",
+		assemblyName, assemblyNumber);
+
+	writtenChars += sprintf_s(assemblyInfo + writtenChars, BUFFER_SIZE - writtenChars, " Version:%i.%i.%i.%i",
+		metadata.usMajorVersion, metadata.usMinorVersion, metadata.usBuildNumber, metadata.usRevisionNumber);
```
Using `sprintf_s` is error-prone.
The more modern way (C++11) is to use `std::ostringstream` or even better (C++20) `std::format`

wilhelma commented on Profiler/CProfilerCallback.cpp:
```diff
@@ -1,488 +1,555 @@
+	}
+
+	if (config.shouldLogAssemblyPaths()) {
+		sprintf_s(assemblyInfo + writtenChars, BUFFER_SIZE - writtenChars, " Path:%S", assemblyPath);
```
again `sprintf_s`

wilhelma commented on Profiler/CProfilerCallback.cpp:
```diff
@@ -1,488 +1,555 @@
+}
+
+int CProfilerCallback::registerAssembly(AssemblyID assemblyId) {
+	int assemblyNumber = assemblyCounter;
+	assemblyCounter++;
```
that's what `postfix++` does :)
```suggestion
	int assemblyNumber = assemblyCounter++;
```

wilhelma commented on Profiler/CProfilerCallback.cpp:
```diff
@@ -1,488 +1,555 @@
+void CProfilerCallback::handleException(std::string context) {
+	Debug::getInstance().logErrorWithStracktrace(context);
+	if (!config.shouldIgnoreExceptions()) {
+		throw;
```
I am not sure why no exception object is used here..
This may result in implementation-defined program termination when there is no exception object to re-throw.

wilhelma commented on Profiler/CProfilerCallback.cpp:
```diff
@@ -1,488 +1,555 @@
+		return 0;
+	}
+
+	BYTE* versionInfo = new BYTE[infoSize];
```
Use `unique_ptr`
```suggestion
	unique_ptr<BYTE[]> versionInfo{new BYTE[infoSize]};
```

wilhelma commented on Profiler/CProfilerCallback.cpp:
```diff
@@ -1,488 +1,555 @@
+
+	VS_FIXEDFILEINFO* fileInfo = NULL;
+	UINT fileInfoLength = 0;
+	if (!VerQueryValueW(versionInfo, L"\\", (void**)&fileInfo, &fileInfoLength)) {
```
```suggestion
	if (!VerQueryValueW(versionInfo.get(), L"\\", (void**)&fileInfo, &fileInfoLength)) {
```

wilhelma commented on Profiler/CProfilerCallback.cpp:
```diff
@@ -1,488 +1,555 @@
+	writtenChars += sprintf_s(buffer + writtenChars, bufferSize - writtenChars, " ProductVersion:%i.%i.%i.%i",
+		version[0], version[1], version[2], version[3]);
+
+	delete versionInfo;
```
Not necessary with unique pointers
```suggestion
```

wilhelma commented on Profiler/CProfilerCallback.cpp:
```diff
@@ -1,488 +1,555 @@
+	}
+}
+
+void CProfilerCallback::onTestEnd(std::string result, std::string duration)
```
```suggestion
void CProfilerCallback::onTestEnd(const std::string& result, const std::string& duration)
```

wilhelma commented on Profiler/log/TraceLog.h:
```diff
@@ -11,7 +11,7 @@
+class TraceLog : public FileLogBase
 {
 public:
 	virtual ~TraceLog() noexcept;
```
```suggestion
        ~TraceLog() override = default;
```

wilhelma commented on Profiler/log/TraceLog.cpp:
```diff
@@ -1,53 +1,55 @@
+#include <regex>
+#include <sstream>
+
+TraceLog::~TraceLog() {
+	// Nothing to do here, destructing is handled in FileLogBase
+}
```
Remove, see header file
```suggestion

```

wilhelma commented on Profiler/log/TraceLog.cpp:
```diff
@@ -1,53 +1,55 @@
+	// Nothing to do here, destructing is handled in FileLogBase
+}
+
+void TraceLog::writeJittedFunctionInfosToLog(std::vector<FunctionInfo>* functions)
```
Generic: Is it intended to use a pointer here? I would assume that this suffices (as the vector is probably not changed)
```suggestion
void TraceLog::writeJittedFunctionInfosToLog(const std::vector<FunctionInfo>& functions)

```

wilhelma commented on Profiler/log/TraceLog.cpp:
```diff
@@ -1,53 +1,55 @@
+	writeFunctionInfosToLog(LOG_KEY_CALLED, functions);
+}
+
+void TraceLog::createLogFile(std::string targetDir) {
```
```suggestion
void TraceLog::createLogFile(const std::string& targetDir) {

```

wilhelma commented on Profiler/log/TraceLog.cpp:
```diff
@@ -1,53 +1,55 @@
+void TraceLog::writeFunctionInfosToLog(const char* key, std::vector<FunctionInfo>* functions) {
+	std::stringstream stream;
+	std::string endLine = "\r\n";
+	for (std::vector<FunctionInfo>::iterator i = functions->begin(); i != functions->end(); i++) {
+		stream << key << '=' << i->assemblyNumber << ':' << i->functionToken << endLine;
+	}
```
Why an iterator?
```suggestion
	for (const FunctionInfo& function : functions) {
		stream << key << '=' << function.assemblyNumber << ':' << function.functionToken << endLine;
	}

```

wilhelma commented on Profiler/log/TraceLog.cpp:
```diff
@@ -77,11 +79,32 @@ void TraceLog::logAssembly(std::string assembly)
+{
+	// Line will look like this:
+	// Test=End:{End Date}:{Result}:{Duration}
+	std::string info = "End:" + getFormattedCurrentTime();
```
Use a different variable name than `info` as it is the name of a method already

wilhelma commented on Profiler/utils/Ipc.cpp:
```diff
@@ -0,0 +1,135 @@
+
+using namespace std::chrono_literals;
+
+#define IPC_TIMEOUT_MS 1000
+#define IPC_BUFFER_SIZE 768
+#define IPC_LINGER 0
+
+// ZMQ strings are not zero-terminated, let's terminate these
+#define IPC_TERMINATE_STRING(buffer, len) buffer[len < IPC_BUFFER_SIZE ? len : IPC_BUFFER_SIZE - 1] = '\0'
```
Don't use macros (`#define`) but `constexpr` variables for type safety (and more)

wilhelma commented on Profiler/utils/Ipc.cpp:
```diff
@@ -0,0 +1,135 @@
+const std::string TEST_START = "start:";
+const std::string TEST_END = "end:";
+
+Ipc::Ipc(Config* config, std::function<void(std::string)> testStartCallback, std::function<void(std::string, std::string)> testEndCallback, std::function<void(std::string)> errorCallback)
```
```suggestion
Ipc::Ipc(Config* config, const std::function<void(std::string)>& testStartCallback, const std::function<void(std::string, std::string)>& testEndCallback, const std::function<void(std::string)>& errorCallback)
```

wilhelma commented on Profiler/utils/Ipc.cpp:
```diff
@@ -0,0 +1,135 @@
+	this->zmqLinger = IPC_LINGER;
+	this->zmqContext = zmq_ctx_new();
+
+	this->handlerThread = new std::thread(&Ipc::handlerThreadLoop, this);
```
This is a memory leak. Use `unique_ptr`

wilhelma commented on Profiler/utils/Ipc.cpp:
```diff
@@ -0,0 +1,135 @@
+
+void Ipc::handlerThreadLoop() {
+	std::string address = "";
+	while (address == "" && !this->shutdown) {
```
```suggestion
	while (address.empty() && !this->shutdown) {
```

wilhelma commented on Profiler/utils/Ipc.cpp:
```diff
@@ -0,0 +1,135 @@
+	std::string address = "";
+	while (address == "" && !this->shutdown) {
+		address = this->request("register:" + std::to_string(GetCurrentProcessId()));
+		if (address == "") {
```
```suggestion
		if (address.empty()) {
```

wilhelma commented on Profiler/utils/Ipc.cpp:
```diff
@@ -0,0 +1,135 @@
+
+Ipc::Ipc(Config* config, std::function<void(std::string)> testStartCallback, std::function<void(std::string, std::string)> testEndCallback, std::function<void(std::string)> errorCallback)
+{
+	this->config = config;
+	this->testStartCallback = testStartCallback;
+	this->testEndCallback = testEndCallback;
+	this->errorCallback = errorCallback;
+	this->zmqTimeout = IPC_TIMEOUT_MS;
+	this->zmqLinger = IPC_LINGER;
+	this->zmqContext = zmq_ctx_new();
```
The preferred way is to use [member initialization lists](https://en.cppreference.com/w/cpp/language/constructor)

wilhelma commented on Profiler/utils/Ipc.cpp:
```diff
@@ -0,0 +1,135 @@
+	}
+
+	while (!this->shutdown) {
+		char buf[IPC_BUFFER_SIZE];
```
use `std::array`

wilhelma commented on Profiler/utils/Ipc.cpp:
```diff
@@ -0,0 +1,135 @@
+	return this->request("testname");
+}
+
+std::string Ipc::request(std::string message)
```
```suggestion
std::string Ipc::request(const std::string& message)
```

wilhelma commented on Profiler/utils/Ipc.cpp:
```diff
@@ -0,0 +1,135 @@
+		return "";
+	}
+
+	char buffer[IPC_BUFFER_SIZE];
```
std::array

wilhelma commented on Profiler/utils/Ipc.h:
```diff
@@ -0,0 +1,32 @@
+public:
+	Ipc(Config* config, std::function<void(std::string)> testStartCallback, std::function<void(std::string, std::string)> testEndCallback, std::function<void(std::string)> errorCallback);
+	~Ipc();
+	std::string getCurrentTestName();
```
missing comment

wilhelma commented on Profiler/utils/Ipc.h:
```diff
@@ -0,0 +1,32 @@
+#include <atomic>
+#include <functional>
+
+class Ipc
```
missing comment

wilhelma commented on Profiler/utils/MethodEnter.h:
```diff
@@ -0,0 +1,30 @@
+/**
+ * Sets the state of test case recording i.e. whether a test case is currently in progress or not.
+ */
+void setTestCaseRecording(bool);
```
Give the comment a name

wilhelma commented on Profiler/utils/MethodEnter.h:
```diff
@@ -0,0 +1,30 @@
+ */
+void setCalledMethodsSet(function_id_set*);
+
+void setCriticalSection(CRITICAL_SECTION*);
```
missing comment

wilhelma commented on Profiler/utils/functionID_set/functionId_set.h:
```diff
@@ -0,0 +1,165 @@
+/// Set that can only contain functionIDs, which are just unsigned ints.
+/// This is based on an array and is a lot faster than the default set implementation of the standard library for this use case.
+/// </summary>
+class function_id_set final
```
Violates typical C++ naming conventions (camel case)

wilhelma commented on Profiler/utils/functionID_set/functionId_set.h:
```diff
@@ -0,0 +1,165 @@
+	unsigned int num_elements = 0;
+	unsigned int max_elements = DEFAULT_SIZE / 2;
+	unsigned int modulo_mask = current_size - 1;
+	FunctionID* set = new FunctionID[DEFAULT_SIZE]{ 0 };
```
unique_ptr

wilhelma replied:
std::array

wilhelma commented on Profiler/utils/functionID_set/functionId_set.h:
```diff
@@ -0,0 +1,165 @@
+
+public:
+
+	~function_id_set() {
```
not necessary with unique_ptr

wilhelma commented on Profiler/utils/functionID_set/functionId_set.h:
```diff
@@ -0,0 +1,165 @@
+		current_size = DEFAULT_SIZE;
+		max_elements = DEFAULT_SIZE / 2;
+		num_elements = 0;
+		set = new FunctionID[DEFAULT_SIZE]{ 0 };
```
memory leak
use `reset` of `unique_ptr`

wilhelma commented on Profiler/utils/functionID_set/functionId_set.h:
```diff
@@ -0,0 +1,165 @@
+	/// <summary>
+	/// True if the set contains FunctionID f, false otherwise.
+	/// </summary>
+	bool contains(FunctionID f) {
```
I don't understand the logic... Is this really so complicated or can it be used by an algorithm from `<algorithm>`?

Commit:
Rework

Raphael-N commented on Profiler/utils/Ipc.cpp:
```diff
@@ -0,0 +1,135 @@
+}
+
+void Ipc::handleMessage(std::string message) {
+	if (message.find(TEST_START) == 0) {
```
I believe, that's only available in C++20.

wilhelma reviewed (approved):
Changes look good. Is the behavior identical?

snowbagoly committed:
Rework Profiler_Test

snowbagoly committed:
Refactor UploadDaemon_Tests and UploadTask

Raphael-N committed:
Fix mocks in tests

Commit:
TS-38029 Rework

Commit:
TS-38029 Rework

Commit:
Merge branch 'master' into tia

snowbagoly committed:
TS-38029 Fix merge errors

Commit:
TS-38029 Fix archive tests

snowbagoly committed:
TS-38029 Fix tests

snowbagoly committed:
Hackfix setup-rsync

snowbagoly committed:
TS-38029 Small fixes
- Double newlines
- Error on parsing empty lines
- Error on auto-property initializers, see TS-46522
- Artifactory upload did not contain report type in path
- Cache never invalidated

snowbagoly committed:
TS-38029 Add changelog

snowbagoly committed:
TS-38029 Fix incorrect docs

Commit:
TS-38029 Rename "No Test" to "Coverage Outside Tests"
